account_financial_report_webkit/models/account_static_balance.py

```diff
@@ -53,119 +61,179 @@ def unlink(self):
+                return {}
 
         logger.debug(
             "Starting calculation of balances for "
-            "periods: %s.", ", ".join(periods.mapped('name')))
+            "periods: %s.", period.name)
 
         start_time = time()
 
+        for journal in journals:
+            balances = self.env['account.static.balance'].search([
+                ('period_id', '=', period.id),
+                ('journal_id', '=', journal.id)
+            ])
+            existing_accounts = balances.mapped('account_id')
+
+            if existing_accounts and existing_accounts != accounts:
+                balances.unlink()
+
+        # Filter out journals that are already calculated
+        existing_journals = self.env['account.static.balance'].search([
+            ('period_id', '=', period.id)
+        ]).mapped('journal_id')
+
+        journals = journals - existing_journals
+
+        if not journals:
+            # Be sure that we have journals left, otherwise it would not be
+            # necessarily to execute a heavy query
+            return {}
+
         # Calculate the values for the periods
         query = """
-            WITH expanded AS (
-              SELECT
-                ap.id AS period_id,
-                aa.id AS account_id,
-                0 AS debit,
-                0 AS credit,
-                0 AS balance,
-                0 AS curr_balance
-              FROM account_account aa
-              CROSS JOIN account_period ap
-              WHERE ap.id IN %(period_ids)s AND aa.id in %(account_ids)s
-              UNION ALL
-              SELECT
-                ap.id AS period_id,
-                aa.id AS account_id,
+            SELECT
+                aml.period_id AS period_id,
+                aml.account_id AS account_id,
+                aml.journal_id AS journal_id,
                 COALESCE(SUM(debit), 0) AS debit,
                 COALESCE(SUM(credit), 0) AS credit,
                 COALESCE(sum(debit), 0) - COALESCE(sum(credit), 0) AS balance,
                 COALESCE(sum(amount_currency), 0) AS curr_balance
               FROM account_move_line aml
-              JOIN account_account aa ON aa.id = account_id
-              JOIN account_period ap ON ap.id = aml.period_id
               JOIN account_move am ON aml.move_id = am.id
-              WHERE ap.id IN %(period_ids)s AND aa.id in %(account_ids)s
-                {}  -- AND am.state == 'posted'
-              GROUP BY ap.id, aa.id)
-            SELECT
-              period_id AS period_id,
-              account_id AS account_id,
-              SUM(debit) AS debit,
-              SUM(credit) AS credit,
-              sum(balance) as balance,
-              sum(curr_balance) as curr_balance
-              FROM expanded
-            GROUP BY period_id, account_id
+                WHERE aml.period_id=%(period_id)s
+                    AND aml.account_id IN %(account_ids)s
+                    AND aml.journal_id IN %(journal_ids)s
```

Comment:
I would suggest to make these clauses conditional. Especially in the case of accounts, the performance might be worse if you pass the list with all existing accounts instead of leaving out the filter.

Reply:
Great idea! Implemented.